test/extensions/filters/network/common/fuzz/network_readfilter_fuzz_test.cc

```diff
@@ -9,41 +9,87 @@
+    random.initializeSeed(seed);
+    return true;
+  }();
+
+  FuzzerProtoType input;
+  protobuf_mutator::ParseTextMessage(data, size, &input);
+  if (random.random() % 100 < config_mutation_probability) {
+    test::extensions::filters::network::FuzzHelperForActions actions;
+    *actions.mutable_actions() = *input.mutable_actions();
+    mutator.Mutate(&actions, max_size);
+    *input.mutable_actions() = *actions.mutable_actions();
+  } else {
+    mutator.Mutate(input.mutable_config(), max_size);
+    ensuredValidFilter(seed, input.mutable_config());
+  }
+  ProtobufMessage::traverseMessage(generator, input, true);
+
+  return protobuf_mutator::SaveMessageAsText(input, data, max_size);
+}
+
+DEFINE_CUSTOM_PROTO_CROSSOVER_IMPL(false, FuzzerProtoType)
+DEFINE_TEST_ONE_PROTO_INPUT_IMPL(false, FuzzerProtoType)
+DEFINE_POST_PROCESS_PROTO_MUTATION_IMPL(FuzzerProtoType)
+static void TestOneProtoInput(const test::extensions::filters::network::FilterFuzzTestCase& input) {
   TestDeprecatedV2Api _deprecated_v2_api;
   ABSL_ATTRIBUTE_UNUSED static PostProcessorRegistration reg = {
       [](test::extensions::filters::network::FilterFuzzTestCase* input, unsigned int seed) {
         // This post-processor mutation is applied only when libprotobuf-mutator
         // calls mutate on an input, and *not* during fuzz target execution.
         // Replaying a corpus through the fuzzer will not be affected by the
         // post-processor mutation.
-
-        // TODO(jianwendong): After extending to cover all the filters, we can use
-        // `Registry::FactoryRegistry<
-        // Server::Configuration::NamedNetworkFilterConfigFactory>::registeredNames()`
-        // to get all the filter names instead of calling `UberFilterFuzzer::filter_names()`.
-        static const auto filter_names = UberFilterFuzzer::filterNames();
-        static const auto factories = Registry::FactoryRegistry<
-            Server::Configuration::NamedNetworkFilterConfigFactory>::factories();
-        // Choose a valid filter name.
-        if (std::find(filter_names.begin(), filter_names.end(), input->config().name()) ==
-            std::end(filter_names)) {
-          absl::string_view filter_name = filter_names[seed % filter_names.size()];
-          if (filter_name != input->config().name()) {
-            // Clear old config, or unpacking non-suitable value may crash.
-            input->mutable_config()->clear_typed_config();
-            input->mutable_config()->set_name(std::string(filter_name));
-          }
-        }
-        // Set the corresponding type_url for Any.
-        auto& factory = factories.at(input->config().name());
-        input->mutable_config()->mutable_typed_config()->set_type_url(
-            absl::StrCat("type.googleapis.com/",
-                         factory->createEmptyConfigProto()->GetDescriptor()->full_name()));
+        ensuredValidFilter(seed, input->mutable_config());
 
         ProtobufMessage::ValidatedInputGenerator generator(
```

Comment:
Maybe I'm missing something, but isn't the generator called already above?

Reply:
Correct, but the cross-over function may also change the input before it is supplied by the post process. But you are super correct here, that it would be duplicate to call the valid input generator here and in the mutate function. I therefore removed it from the mutate function. This post processor then will ensure that the input is correct.